README.md
# apigee-tooling

Google APIGEE X administration tasks helpers.

The scripts and programs in this report probably have more limitations than [Apigee Sackmesser](https://github.com/apigee/devrel/tree/main/tools/apigee-sackmesser) tool, provided by Google Apigee team, but it was required for my job.

However, it may help some folks.

> **_DISCLAIMER_**
>
> Use these scripts at your own risks. Developer can not be considered as responsible in case of KVM or entries deletion or modification in case of script bug or misusage.


bundle-compare/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.1 - 2022-08-02]

### Added

- Script and git context creation


lib/logutils.sh
#!/bin/bash
# shellcheck disable=SC2154
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# <http://www.apache.org/licenses/LICENSE-2.0>
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
RED='\033[0;31m'
ORANGE='\033[0;33m'
BLUE='\033[0;36m'
NOCOL='\033[0m'

NOOUTPUT=''

logdebug() {
    if [ "$debug" = "T" ]; then
        log "DEBUG" "$@" "$BLUE"
    fi
}

loginfo() {
        log "INFO" "$@" "$NOCOL"
}

logwarn() {
        log "WARN" "$@" "$ORANGE"
}

logerror() {
        log "ERROR" "$@" "$RED"
}

logfatal() {
    log "FATAL" "$@" "$RED"
}

log() {
    echo -e "${3}[$1] $2${NOCOL}" 1>&2;
}

lib/apigee-common.sh
checkOrganization () {

    local ORG=$1

    loginfo "ORG - Checking organization $ORG"

    URI="https://apigee.googleapis.com/v1/organizations/$ORG"
    RESPONSE=$(curl -s -w "%{http_code}" -X GET -H "Authorization: Bearer $AUTH" $URI)
    RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
    CONTENT=$(sed '$ d' <<< "$RESPONSE")

    logdebug "ORG - URI: $URI"
    logdebug "ORG - RETURN_CODE: $RETURN_CODE"
    
    if [ $RETURN_CODE != "200" ]; then
        logdebug "ORG - CONTENT: $CONTENT"
        logfatal "ORG - Check of $ORG organization failed. Exiting."
        exit 5
    fi
}


checkEnvironment () {
    local ORG=$1
    local ENV=$2

    loginfo "ENV - Checking environment $ENV"

    URI="https://apigee.googleapis.com/v1/organizations/$ORG/environments/$ENV"
    RESPONSE=$(curl -s -w "%{http_code}" -X GET -H "Authorization: Bearer $AUTH" $URI)
    RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
    CONTENT=$(sed '$ d' <<< "$RESPONSE")

    logdebug "ENV - URI: $URI"
    logdebug "ENV - RETURN_CODE: $RETURN_CODE"
    
    if [ $RETURN_CODE != "200" ]; then
        logdebug "ENVRG - CONTENT: $CONTENT"
        logfatal "ENV - Check of $ORG/$ENV environment failed. Exiting."
        exit 5
    fi
}

getAuth () {
    if [ -z ${AUTH} ]; then
        loginfo "AUTH - Generating a new token"
        AUTH=$(gcloud auth print-access-token)
    else
        loginfo "AUTH - AUTH variable does exist and will be used to authenticate API calls."
    fi
    logdebug "${AUTH:0:15}*****"
}

getEnvironmentDeployments () { 
    local ORG=$1
    local ENV=$2

    loginfo "ENVDEPL - Get deployments on organization/$ORG/environments/$ENV"

    URI="https://apigee.googleapis.com/v1/organizations/$ORG/environments/$ENV/deployments"

    RESPONSE=$(curl -s -w "\n%{http_code}" -X GET -H "Authorization: Bearer $AUTH" $URI)
    RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
    CONTENT=$(sed '$ d' <<< "$RESPONSE")

    logdebug "RETURN CODE : $RETURN_CODE"
}

kvm-terraforming/kvm-terraforming.sh
#!/bin/bash

## ######################################
## kvm-terraform.sh
##
## Version: 0.1
## Date: 2022-06-24
###
## Exit codes:
##  1: Wrong number of script argument (should be 1)
##  2: Argument is not an existing file 
##  3: Dependencies are not satisfied
##  4: Missing mandatory data on the argument JSON file 
##  5: Error while getting Google Cloud credentials 
##  6: Error while creating the kvm
##  7: Error while creating keys
## ######################################

# Small library for better log display
SCRIPT_ROOT=$( (cd "$(dirname "$0")" && cd .. && pwd ))
source "$SCRIPT_ROOT/lib/logutils.sh"
source "$SCRIPT_ROOT/lib/apigee-common.sh"

## FUNCTIONS ##########
upsert_kvm () {

    ## ########
    ## To create the MAP if it does not exist
    ## 
    ## Params
    ##   $1: Organization name
    ##   $2: Environment name 
    ##   $3: Map name name 
    ## ########

    loginfo "Processing KEYMAP $1 // $2 // $3"

    URI="https://apigee.googleapis.com/v1/organizations/$1/environments/$2/keyvaluemaps/$3"
    RETURN_CODE=$(curl -XGET -s -o /dev/null -w "%{http_code}" -H "Authorization: Bearer $AUTH" $URI/entries)
    if [ $RETURN_CODE -eq "200" ]
    then
      loginfo "The KVM does already exist."
    else
      loginfo "The KVM does NOT exist. Requesting creation."
      URI="https://apigee.googleapis.com/v1/organizations/$1/environments/$2/keyvaluemaps"

      RESPONSE=$(curl -s -w "%{http_code}" -XPOST -H "Content-Type: application/json" -H "Authorization: Bearer $AUTH" -d '{"name":"'$3'","encrypted":true}' $URI)
      RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
      CONTENT=$(sed '$ d' <<< "$RESPONSE")

      if [ $RETURN_CODE -eq "201" ]
      then
        loginfo "Keyvaluemap $3 successfully created."
      else
        logfatal "Something FAILED."
        logfatal "$CONTENT"
        logfatal "Stopping..."
        exit 6
      fi

      #loginfo "Command to delete the KVM: curl -XDELETE -H 'Authorization: Bearer xxxx' https://apigee.googleapis.com/v1/organizations/$1/environments/$2/keyvaluemaps/$3"
    fi
}

upsert_keys () {

    ## ########
    ## To delete/create the keys
    ## 
    ## Params
    ##   $1: Organization name
    ##   $2: Environment name 
    ##   $3: Map name name
    ##   $4: the keys JSON array
    ## ########

    loginfo "Processing KEYS for $1 // $2 // $3"
    
    # Broswing every key
    jq -c -r '.[] | select (.!=null)' <<< "$4" | while read KEY; do

      # Extractin keyname and keyvalue
      KEYNAME=$(jq -c -r '.name' <<< "$KEY")
      KEYVALUE=$(jq -c -r '.value' <<< "$KEY")
      loginfo "Creation key '$KEYNAME' with value '${KEYVALUE:0:5}******' (hidden for confidentiality)"

      # DELETE FIRST
      URI="https://apigee.googleapis.com/v1/organizations/$1/environments/$2/keyvaluemaps/$3/entries/$KEYNAME"
      RESPONSE=$(curl -s -w "%{http_code}" -X DELETE -H "Content-Type: application/json" -H "Authorization: Bearer $AUTH" $URI)
      RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
      CONTENT=$(sed '$ d' <<< "$RESPONSE")
      
      if [ $RETURN_CODE -eq "200" ]; then
        loginfo "Keys '$KEYNAME': deleted"
      fi

      # CREATE THEN
      URI="https://apigee.googleapis.com/v1/organizations/$1/environments/$2/keyvaluemaps/$3/entries"
      RESPONSE=$(curl -s -w "%{http_code}" -X POST -H "Content-Type: application/json" -H "Authorization: Bearer $AUTH" -d '{"name":"'$KEYNAME'","value":"'$KEYVALUE'"}' $URI)
      RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
      CONTENT=$(sed '$ d' <<< "$RESPONSE")
      
      if [ $RETURN_CODE -eq "201" ]; then
        loginfo "Keys '$KEYNAME': created"
      else
        # Dmaned, something went wrong.
        logfatal "Error while creating key $KEYNAME"
        logfatal "$CONTENT"
        exit 7
      fi
    done
}





## SCRIPT ##########

# Check number of argument
if [ $# -ne 1 ]
then
    logerror "Wrong number of arguments"
    exit 1
fi

# Check if argument is a file
if [ ! -f $1 ]
then
    logerror "The argument must be an existing file"
    exit 2
fi



# Check dependencies
loginfo "Checking dependencies"
for dependency in jq gcloud
do
  if ! [ -x "$(command -v $dependency)" ]; then
    >&2 logfatal "Required command is not on your PATH: $dependency."
    >&2 logfatal "Please install it before you continue."
    exit 3
  fi
done


# Retrieves data from parameter file
KVMNAME=$(jq -c -r '.name | select (.!=null)' $1)
ORG=$(jq -c -r '.organization | select (.!=null)' $1)
ENV=$(jq -c -r '.environment | select (.!=null)' $1)
KEYS=$(jq -c -r '.keys' $1)

if [ -z ${KVMNAME:+x} ]; then
  logfatal "KVM Name not found"
  stop=1
fi
if [ -z ${ORG:+x} ]; then
  logfatal "ORG not found"
  stop=1
fi
if [ -z ${ENV:+x} ]; then
  logfatal "ENV not found"
  stop=1
fi
if [ ! -z ${stop:+x} ]; then
  logfatal "Missing mandatory information. Stopping here!"
  exit 4
fi


# GCLoud CREDS
loginfo "Getting GCloud credentials"
AUTH=$(gcloud auth print-access-token)

if [ -z ${AUTH:+x} ]; then
  logfatal "Error while getting creds. Stopping."
  exit 5
fi

# Check if ORG is OK
checkOrganization $ORG
# Check if ENV is OK
checkEnvironment $ORG $ENV


# Preliminaries are good. Let's start
# First, the MAP
upsert_kvm $ORG $ENV $KVMNAME 
# Second, the KEYS
upsert_keys $ORG $ENV $KVMNAME $KEYS

loginfo "Success. End of process"

deployments-status/deployments-status.sh
#!/bin/bash
# Goal: generate a file with status (=name and revision) deployed on every environment of an org (+ hash ?)



## NOTE : this script only output DEBUG log, as the "normal" behavior will be to display directly the result
##        Reason is CICD: if "normal" result can be parsed or redirected into file, that could help for CICD/automation
SCRIPT_ROOT=$( (cd "$(dirname "$0")" && cd .. && pwd ))
source "$SCRIPT_ROOT/lib/logutils.sh"
source "$SCRIPT_ROOT/lib/apigee-common.sh"

usage () {
    echo "Usage: deployments-status.sh -c <configFile.json> [-d] [-h]"
    echo ""
    echo "  -c <configFile.json>    MANDATORY   : the fullpath of config file that list org/env to browse"
    echo "  -d                      OPTIONAL    : Activate the debug log mode"
    echo "  -h                      OPTIONAL    : Display THIS message"
    echo ""
    echo ""
    echo "Script display logs on STDERR, and useful result on STDOUT."
    echo ""
    echo "To get in the console only the useful result, you can use: deployment-status.sh -c config.json 2> /dev/null"
    echo ""
    echo "To display a sorted version (more readability of identical hashes): deployment-status.sh -c config.json 2> /dev/null | sort -k6"
    echo ""
}

getHash () {
    local ORG=$1
    local ENV=$2
    local PROXYNAME=$3
    local PROXYREVISION=$4
    local PROXYDATE=$5

    logdebug "HASH - Inputs: $ORG/$ENV/$PROXYNAME/$PROXYREVISION"
    FILENAME=$(mktemp)
    logdebug "HASH - Target filename: $FILENAME"

    loginfo "HASH - Getting BUNDLE for $ORG // $PROXYNAME // $PROXYREVISION"
    URI="https://apigee.googleapis.com/v1/organizations/$ORG/apis/$PROXYNAME/revisions/$PROXYREVISION?format=bundle"

    RESPONSE=$(curl -s -o $FILENAME -w "%{http_code}" -X GET -H "Authorization: Bearer $AUTH" $URI)
    RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
    CONTENT=$(sed '$ d' <<< "$RESPONSE")

    logdebug "HASH - URI: $URI"
    logdebug "HASH - RETURN_CODE: $RETURN_CODE"
    
    if [ $RETURN_CODE != "200" ]; then
        logfatal "BUNDLE - Downloading the ZIP bundle failed. Exiting."
        exit 8
    fi

    logdebug "$(unzip -vqql "$FILENAME" | awk '{printf("%s %s\n",$7,$8)}' | sort -k2 | tail -n +3)"

    HASH=($(md5sum <<< $(unzip -vqql "$FILENAME" | awk '{printf("%s %s\n",$7,$8)}' | sort -k2 | tail -n +3)))

    if [ $? != 0 ]; then
        logfatal "HASH - Error while calculating the HASH. Exiting."
        exit 9
    fi
    displayProxyRevisionHash $ORG $ENV $PROXYNAME $PROXYREVISION $PROXYDATE $HASH
    rm $FILENAME
}

displayProxyRevisionHash () {
    local ORG=$1
    local ENV=$2
    local PROXYNAME=$3
    local PROXYREVISION=$4
    local PROXYDATE=$5
    local HASH=$6

    FORMATROW="%-20s %-8s %-30s %-4s %-20s %-32s\n"
    printf "$FORMATROW" "$ORG" "$ENV" "$PROXYNAME" "$PROXYREVISION" "$PROXYDATE" "$HASH"
}

getDeployments () {
    logdebug "GETDEPL - organizations/$ORG/environments/$ENV"
    getEnvironmentDeployments $ORG $ENV
    DEPL=$(jq -r '.deployments[] | .apiProxy + "=" +  .revision + "=" + .deployStartTime' <<< $CONTENT)
    
    while IFS= read -r oneDepl
    do
        PROXYNAME=$(echo $oneDepl | cut -f1 -d=)
        PROXYREVISION=$(echo $oneDepl | cut -f2 -d=)
        PROXYDATE=$(echo $oneDepl | cut -f3 -d=)
        PROXYDATE=$(date -d @${PROXYDATE::-3} +'%FT%T')
        getHash $ORG $ENV $PROXYNAME $PROXYREVISION $PROXYDATE
    done <<< "$DEPL"
}

parseFile () {

    logdebug "PARSE - Browsing config file setup"
    INSTANCES=$(jq -c -r '.instances' $FULLPATHFILE)
    local L=$(jq -c -r '. | length' <<< $INSTANCES)

    if [  $L -lt 1 ]; then
        logfatal "PARSE - No instance found in the file."
        exit 7
    else
        logdebug "PARSE - Config file has $L organizations"
    fi 
    
    ORGS=$(jq -c -r '.[].organization' <<< $INSTANCES)

    loginfo "PARSE - Getting authentication gcloud token"
    getAuth




    for ORG in $ORGS
    do
        logdebug "PARSE - Check existence of organizations/$ORG"
        checkOrganization $ORG

        ENVS=$(jq --arg o "$ORG" -c -r '.[]| select(.organization | contains($o)) | .environments[]' <<< $INSTANCES)

        for ENV in $ENVS
        do
            logdebug "PARSE - Check existence of organization/$ORG/environments/$ENV"
            checkEnvironment $ORG $ENV
            getDeployments $ORG $ENV 
        done
    done

}

checkPrereq () {

    # Checking mandatory arguments
    if [ -z "${CONFIG}" ]; then
        logfatal "PREREQ - Missing confif file argument (-c)"
        usage
        exit 5
    else
        logdebug "PREREQ - Config file argument: $CONFIG"
    fi

    # Checking if CONFIG is really a file
    FULLPATHFILE=$(readlink -m $CONFIG)
    logdebug "PREREQ - Full path: $FULLPATHFILE"
    if ! [ -f $FULLPATHFILE ]; then
        logfatal "PREREQ - The argument ($FULLPATHFILE) is not a file, or not found"
        exit 6
    else
        logdebug "PREREQ - The argument ($FULLPATHFILE) seems good."
    fi

    # Checking commands prerequisites
    loginfo "PREREQ - Checking dependencies"
    for dependency in jq gcloud md5sum tail curl sed readlink unzip awk sort tail
        do
        logdebug "PREREQ - Checking for command/tool $dependency"
        if ! [ -x "$(command -v $dependency)" ]; then
            >&2 logfatal "PREREQ - Required command is not on your PATH: $dependency."
            >&2 logfatal "PREREQ - Please install it before you continue."
            exit 3
        fi
    done       
}



getDeployements () {
    getAuth

    URI="https://apigee.googleapis.com/v1/organizations/apim-staging-350613/environments/dev/deployments"
    RESPONSE=$(curl -s -w "%{http_code}" -X GET -H "Authorization: Bearer $AUTH" $URI)
    RETURN_CODE=$(tail -n1 <<< "$RESPONSE")
    CONTENT=$(sed '$ d' <<< "$RESPONSE")

    DEPLOYMENTS=$(jq -r -c '.deployments[]' <<< $CONTENT)

    for DEPLOYMENT in ${DEPLOYMENTS[@]}; do
        PROXYNAME=$(jq -r -c '.apiProxy' <<< $DEPLOYMENT)   
        REVISION=$(jq -r -c '.revision' <<< $DEPLOYMENT)   
       
        URI="https://apigee.googleapis.com/v1/organizations/apim-staging-350613/apis/$PROXYNAME/revisions/$REVISION"

        RESPONSE=$(curl -s -o /tmp/$PROXYNAME-$REVISION.zip -X GET -H "Authorization: Bearer $AUTH" $URI)
        
        HASHRESULT=($(md5sum /tmp/$PROXYNAME-$REVISION.zip))

        logdebug "Delete temp fil /tmp/$PROXYNAME-$REVISION.zip"
        rm /tmp/$PROXYNAME-$REVISION.zip

        DEPLOYMENT=$(jq --arg v $HASHRESULT '. += {"md5hash": $v}' <<< $DEPLOYMENT)
    done
}

while getopts "c:dh" option; do
    case "${option}" in
        h)
            usage
            exit 2
            ;;
        c)
            CONFIG=${OPTARG}
            loginfo "MAIN - Config file input: $CONFIG"
            ;;
        d) 
            # DEBUG mode activated
            debug="T"
            logdebug "DEBUG mode activated. "
            ;;
        *)
            usage
            ;;
    esac
done
shift $((OPTIND-1))



checkPrereq
parseFile $CONFIG





bundle-compare/bundle-compare.sh
#!/bin/bash

## ######################################
## bundle-compare.sh
##
## Version: 0.1
## Date: 2022-08-02
###
## Exit codes:
##  0: Bundles are the same (excluding "header" zip file)
##  1: Bundles are NOT the same
##  2: Wrong number of script arguments
##  3: Argument is not an existing file 
##  4: Dependencies are not satisfied
## ######################################

SCRIPT_ROOT=$( (cd "$(dirname "$0")" && cd .. && pwd ))
source "$SCRIPT_ROOT/lib/logutils.sh"

## Inspired from Stackoverflow answer: https://stackoverflow.com/a/44198080

# Test if arguments are there
if [ -z $1 ] || [ -z $2 ]; then
	logfatal "Missing arguments"
    loginfo "Usage: $0 bundle1.zip bundle2.zip"
	exit 2
fi

# Test if arguments are existing files
if ! [ -f $1 ] || ! [ -f $2 ]; then
	logfatal "Argument files not found or are not files"
	exit 3
fi

# Quickly checks for dependencies
for dependency in unzip awk sort tail
do
  if ! [ -x "$(command -v $dependency)" ]; then
    >&2 logfatal "Required command is not on your PATH: $dependency."
    >&2 logfatal "Please install it before you continue."
    exit 4
  fi
done

# Prepare awk format
A='{printf("%8sB %s %s\n",$1,$7,$8)}'

# Run the diff
diffRes=$(diff <(unzip -vqql "$1" | awk "$A" | sort -k3 | tail -n +3) <(unzip -vqql "$2" | awk "$A" | sort -k3 | tail -n +3);)

# Conclusions
if [ $? = 0 ]; then
	exit 0
else
	echo $diffRes
	exit 1
fi
